src/k8s/K8sWorkerController.cpp

```diff
@@ -13,66 +13,88 @@ limitations under the License.
 
 #include "K8sWorkerController.h"
 
-#include <stdexcept>
-#include <utility>
 #include <stdlib.h>
 
-#include "../util/Conts.h"
-#include "../util/logger/Logger.h"
-#include "../server/JasmineGraphServer.h"
+#include <stdexcept>
+#include <utility>
 
 Logger controller_logger;
 
+std::vector<JasmineGraphServer::worker> K8sWorkerController::workerList = {};
+static int TIME_OUT = 900;
+static std::vector<int> activeWorkerIds = {};
+std::mutex workerIdMutex;
+std::mutex k8sSpawnMutex;
+static volatile int nextWorkerId = 0;
+
+static inline int getNextWorkerId(int count) {
+    const std::lock_guard<std::mutex> lock(workerIdMutex);
+    int returnId = nextWorkerId;
+    nextWorkerId += count;
+    return returnId;
+}
+
 K8sWorkerController::K8sWorkerController(std::string masterIp, int numberOfWorkers, SQLiteDBInterface *metadb) {
     this->masterIp = std::move(masterIp);
-    this->numberOfWorkers = numberOfWorkers;
+    this->numberOfWorkers = 0;
+    apiClient_setupGlobalEnv();
     this->interface = new K8sInterface();
     this->metadb = *metadb;
+}
 
-    // Delete all hosts
-    metadb->runUpdate("DELETE FROM host");
+K8sWorkerController::~K8sWorkerController() {
+    delete this->interface;
+    apiClient_unsetupGlobalEnv();
+}
 
-    v1_node_list_t *nodeList = interface->getNodes();
-    listEntry_t *listEntry = NULL;
-    v1_node_t *node = NULL;
+static K8sWorkerController *instance = nullptr;
 
-    std::string query;
-    list_ForEach(listEntry, nodeList->items) {
-        node = static_cast<v1_node_t *>(listEntry->data);
-        query = "INSERT INTO host (name, ip) VALUES ('" + std::string(node->metadata->name) + "', '')";
-        metadb->runInsert(query);
+K8sWorkerController *K8sWorkerController::getInstance() {
+    if (instance == nullptr) {
+        throw std::runtime_error("K8sWorkerController is not instantiated");
     }
+    return instance;
+}
 
-    query = "SELECT idhost, name FROM host";
-    for (auto entry : metadb->runSelect(query)) {
-        nodes[entry.at(1).second] = atoi(entry.at(0).second.c_str());
-    }
+K8sWorkerController *K8sWorkerController::getInstance(std::string masterIp, int numberOfWorkers,
+                                                      SQLiteDBInterface *metadb) {
+    // TODO(thevindu-w): synchronize
```

Comment:
Can we fix this TODO message now itself?

Reply:
Done

Reply:
Ok